### PR TITLE
Fix issues when passing a vararg for non-rest params

### LINKED ISSUE #24281: Taint analyzer crashes when passing a rest arg to a function with no rest param
**Description:**
$title.

**Steps to reproduce:**
```ballerina
import ballerina/io;

public function main() {
    [int, string] details = [1234, "IT"];
    Employee e = createEmployee("Jo", ...details);

    io:println(e.name);
    io:println(e.id);
    io:println(e.department);
}

function createEmployee(string name, int id, string department) returns Employee => {
    name, id, department
};

type Employee record {|
    string name;
    int id;
    string department;
|};
```

Crashes at the moment, but should pass and run.
```
[2020-06-18 00:52:41,997] SEVERE {b7a.log.crash} - Index: 3, Size: 3 
java.lang.IndexOutOfBoundsException: Index: 3, Size: 3
	at java.util.ArrayList.rangeCheck(ArrayList.java:653)
	at java.util.ArrayList.get(ArrayList.java:429)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.updateArgTaintedStatus(TaintAnalyzer.java:2676)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeInvocation(TaintAnalyzer.java:2432)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:1077)
	at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation.accept(BLangInvocation.java:115)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeNode(TaintAnalyzer.java:1661)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:505)
	at org.wso2.ballerinalang.compiler.tree.BLangSimpleVariable.accept(BLangSimpleVariable.java:53)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:590)
	at org.wso2.ballerinalang.compiler.tree.statements.BLangSimpleVariableDef.accept(BLangSimpleVariableDef.java:46)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeNode(TaintAnalyzer.java:1661)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:414)
	at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:57)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeNode(TaintAnalyzer.java:1661)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeReturnTaintedStatus(TaintAnalyzer.java:2091)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeReturnTaintedStatus(TaintAnalyzer.java:2012)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyzeAllParamsUntaintedReturnTaintedStatus(TaintAnalyzer.java:1948)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visitInvokable(TaintAnalyzer.java:1868)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:395)
	at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:73)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.resolveBlockedInvokable(TaintAnalyzer.java:2762)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyze(TaintAnalyzer.java:297)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.visit(TaintAnalyzer.java:277)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:161)
	at org.wso2.ballerinalang.compiler.semantics.analyzer.TaintAnalyzer.analyze(TaintAnalyzer.java:259)
	at org.wso2.ballerinalang.compiler.CompilerDriver.taintAnalyze(CompilerDriver.java:338)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:276)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:248)
	at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:140)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:166)
	at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:174)
	at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:91)
	at org.wso2.ballerinalang.compiler.Compiler.build(Compiler.java:99)
	at org.ballerinalang.packerina.task.CompileTask.execute(CompileTask.java:52)
	at org.ballerinalang.packerina.TaskExecutor.executeTasks(TaskExecutor.java:38)
	at org.ballerinalang.packerina.cmd.RunCommand.execute(RunCommand.java:310)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.tool.Main.main(Main.java:57)

```

### LINKED ISSUE #24348: Invalid type set for the rest arg when the vararg provides for required/defualtable params too
**Description:**
$title.

This is because we use slice internally, which will create an array with all the member types of the original list as the element type.

Need to rewrite this logic without using slice.

**Steps to reproduce:**
```ballerina
public function main() {
    [int, string...] x = [1, "foo"];
    foo(true, ...x);
}

function foo(boolean b, int i, string... s) {
    record {
        string[] s;
    } rec = {s}; // error: {ballerina/lang.map}InherentTypeViolation message=invalid value for record field 's': expected value of type 'string[]', found '(int|string)[1]'
                   // at xyz:foo(xyz.bal:9)
                   // xyz:main(xyz.bal:3)
}
```

**Affected Versions:**
Preview1-M6

### PR DESCRIPTION
**Pull request by MaryamZi**
## Purpose
$title.

Fixes #24348 - Invalid type set for the rest arg when the vararg provides for required/defualtable params too 
Fixes #24281 - Taint analyzer crashes when passing a rest arg to a function with no rest param

## Remarks
https://github.com/ballerina-platform/ballerina-lang/issues/24415
https://github.com/ballerina-platform/ballerina-lang/issues/23307#issuecomment-647352681

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 1]
```diff
@@ -5454,21 +5454,72 @@ private void reorderArguments(BLangInvocation iExpr) {
...
-            // if there is a rest param.
+            // If there's no rest param, the vararg only provided for required/defaultable params.
             if (invokableSymbol.restParam == null) {
+                restArgs.remove(0);
                 return;
             }
 
+            // If there is a rest param, the vararg could provide for the rest param too.
+            // Create a new array with just the members of the original vararg specified for the rest param.
+            BLangRestArgsExpression restArgsExpression = (BLangRestArgsExpression) restArgs.remove(0);
```
Comment (pubudu91) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:5465]: Shall we add a comment about what's being desugared into what?

**Comment by pubudu91** [comment]
The following crashes: 
```ballerina
public function main() {
    [int, (string|int)...] x = [1, 10, "foo"];
    foo(true, ...x);
}

function foo(boolean b, int i, string... s) {
}
```
This should've been caught at compile time right?

**Comment by MaryamZi** [comment]
> The following crashes:
> 
> ```ballerina
> public function main() {
>     [int, (string|int)...] x = [1, 10, "foo"];
>     foo(true, ...x);
> }
> 
> function foo(boolean b, int i, string... s) {
> }
> ```
> 
> This should've been caught at compile time right?

This is due to a bug in tuple type-checking.

How we type-check this is by first creating an expected tuple type based on the remaining non-rest param types and the rest param (if available), which in this case will be `[int, string...]`. We then use this new type as the expected type for the rest-arg expression. But tuple type-checking seems to be broken when the source tuple has a rest descriptor.

Neither of the following `y` and `z` assignments fail.

```ballerina
public function main() {
    [int, (string|int)...] x = [1, 10, "foo"];

    [int, string...] y = x; // no error!
    [int] z = x; // no error!
}
```

Created https://github.com/ballerina-platform/ballerina-lang/issues/24764 to track this.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 2]
```diff
@@ -2445,6 +2473,28 @@ private void analyzeInvocation(BLangInvocation invocationExpr) {
         getCurrentAnalysisState().taintedStatus = returnTaintedStatus;
     }
 
+    private boolean restoreTableIfIgnored(BInvokableSymbol invokableSymbol, Map<Integer, TaintRecord> origTaintTable,
+     TaintedStatus argumentTaintReturnValue) {
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java:2477]: Is this indentation correct here? I think we can change this latter

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 3]
```diff
@@ -2445,6 +2473,28 @@ private void analyzeInvocation(BLangInvocation invocationExpr) {
...
+        if (argumentTaintReturnValue != TaintedStatus.IGNORED) {
+            return false;
+        }
+        invokableSymbol.taintTable = origTaintTable; // restore taint table.
+        return true;
+    }
+
+    private TaintedStatus updateAndGetTaintedStatus(BLangInvocation invocationExpr, TaintedStatus returnTaintedStatus,
```
Comment (rdhananjaya) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java:2485]: Do we need this param `returnTaintedStatus`

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java:2485]: If `argumentTaintReturnValue == TaintedStatus.TAINTED` evaluates to false, we seem to be using this original `returnTaintedStatus`, so I guess we still need it?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java (hunk 2 above)
Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java:2477]: Ack, will fix this.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java (hunk 3 above)
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java:2485]: ack

**Pull request merged**